lib/hmmlearn/base.py

```diff
@@ -494,60 +490,31 @@ def fit(self, X, lengths=None):
             Returns self.
         """
         X = check_array(X)
-        self._init(X)
+        self._init(X, lengths)
         self._check()
-
         self.monitor_._reset()
 
-        impl = {
-            "scaling": self._fit_scaling,
-            "log": self._fit_log,
-        }[self.implementation]
         for iter in range(self.n_iter):
-            stats = self._initialize_sufficient_statistics()
-            curr_log_prob = 0
-            for sub_X in _utils.split_X_lengths(X, lengths):
-                lattice, log_prob, posteriors, fwdlattice, bwdlattice = \
-                        impl(sub_X)
-                # Derived HMM classes will implement the following method to
-                # update their probability distributions, so keep
-                # a single call to this method for simplicity.
-                self._accumulate_sufficient_statistics(
-                    stats, sub_X, lattice, posteriors, fwdlattice,
-                    bwdlattice)
-                curr_log_prob += log_prob
+            stats, curr_logprob = self._do_estep(X, lengths)
+
+            # Compute lower bound before updating model parameters
+            lower_bound = self._compute_lower_bound(curr_logprob)
 
             # XXX must be before convergence check, because otherwise
             #     there won't be any updates for the case ``n_iter=1``.
             self._do_mstep(stats)
-
-            self.monitor_.report(curr_log_prob)
+            self.monitor_.report(lower_bound)
             if self.monitor_.converged:
                 break
-
-        if (self.transmat_.sum(axis=1) == 0).any():
-            _log.warning("Some rows of transmat_ have zero sum because no "
-                         "transition from the state was ever observed.")
-
         return self
 
     def _fit_scaling(self, X):
-        frameprob = self._compute_likelihood(X)
-        log_prob, fwdlattice, scaling_factors = _hmmc.forward_scaling(
-            self.startprob_, self.transmat_, frameprob)
-        bwdlattice =  _hmmc.backward_scaling(
-            self.startprob_, self.transmat_, frameprob, scaling_factors)
-        posteriors = self._compute_posteriors_scaling(fwdlattice, bwdlattice)
-        return frameprob, log_prob, posteriors, fwdlattice, bwdlattice
+        """
```

Comment:
`raise NotImplementedError("Must be overridden in subclass")` rather than an empty docstring? ditto below.